### ISSUE TITLE
The precision of sample_dpmpp_2m_sde and sample_dpmpp_2m_sde_gpu is not equal

### ISSUE THREAD

**Issue by caijizhuo** [comfyanonymous/ComfyUI#7624]
### Expected Behavior

When I use these two different sampler method with the same seed, it should generate the same prictures. However, the pictures are different with huge difference.
In these two sampler functions, the only difference is BatchedBrownianTree. I am not sure if it is the bug of torchsde. I found similar issues mentioned: https://github.com/comfyanonymous/ComfyUI/issues/6474.
Finally I found the precision bug in BatchedBrownianTree. You can reproduce it by the codes below.

### Actual Behavior

The two output tensor of BatchedBrownianTree(GPU and CPU) are the same.

### Steps to Reproduce

test.py:
```python
import torch
import torchsde

class BatchedBrownianTree:
    """A wrapper around torchsde.BrownianTree that enables batches of entropy."""

    def __init__(self, x, t0, t1, seed=None, **kwargs):
        self.cpu_tree = True
        if "cpu" in kwargs:
            self.cpu_tree = kwargs.pop("cpu")
        t0, t1, self.sign = self.sort(t0, t1)
        w0 = kwargs.get('w0', torch.zeros_like(x))
        if seed is None:
            seed = torch.randint(0, 2 ** 63 - 1, []).item()
        self.batched = True
        try:
            assert len(seed) == x.shape[0]
            w0 = w0[0]
        except TypeError:
            seed = [seed]
            self.batched = False
        if self.cpu_tree:
            self.trees = [torchsde.BrownianTree(t0.cpu(), w0.cpu(), t1.cpu(), entropy=s, **kwargs) for s in seed]
        else:
            self.trees = [torchsde.BrownianTree(t0, w0, t1, entropy=s, **kwargs) for s in seed]

    @staticmethod
    def sort(a, b):
        return (a, b, 1) if a < b else (b, a, -1)

    def __call__(self, t0, t1):
        t0, t1, sign = self.sort(t0, t1)
        if self.cpu_tree:
            w = torch.stack([tree(t0.cpu().float(), t1.cpu().float()).to(t0.dtype).to(t0.device) for tree in self.trees]) * (self.sign * sign)
        else:
            w = torch.stack([tree(t0, t1) for tree in self.trees]) * (self.sign * sign)

        return w if self.batched else w[0]
    

# this case is simulating ComfyUI/comfy/k_diffusion/sampling.py:142 (BrownianTreeNoiseSampler)
def test():
    x = torch.tensor([[-11.9037, -13.4259,   1.2816,   5.4348,  -8.2755],
                    [ 17.8416, -11.5165,   4.0193,   0.5852, -14.6976],
                    [-18.8986, -30.2355,   6.2576,   5.7680,   0.2296],
                    [ 22.4932, -17.4192, -12.8944,  13.2802,  -6.3093]], dtype=torch.float32, device="cuda")
    t0 = torch.tensor(0.0292, dtype=torch.float32, device="cuda")
    t1 = torch.tensor(14.6146, dtype=torch.float32, device="cuda")
    seed = 547283076232441

    tree0 = BatchedBrownianTree(x, t0, t1, seed, cpu=True) # cpu tree
    tree1 = BatchedBrownianTree(x, t0, t1, seed, cpu=False) # gpu tree


    t0_ = torch.tensor(14.6146, device="cuda", dtype=torch.float32)
    t1_ = torch.tensor(12.2830, device="cuda", dtype=torch.float32)


    out0 = tree0(t0_, t1_) / (t1_ - t0_).abs().sqrt() # cpu result
    out1 = tree1(t0_, t1_) / (t1_ - t0_).abs().sqrt() # gpu result


    print(out0)
    print("=======")
    print(out1)

    print(out0 == out1)

test()
```
just run it in terminal
```bash
python test.py
```

### Debug Logs

```powershell
root@59561615c495:/workspace# python test.py 
/opt/conda/lib/python3.11/site-packages/torchsde/_brownian/brownian_interval.py:608: UserWarning: Should have tb<=t1 but got tb=14.61460018157959 and t1=14.6146.
  warnings.warn(f"Should have {tb_name}<=t1 but got {tb_name}={tb} and t1={self._end}.")
tensor([[ 0.5589, -1.2761,  0.7936,  0.0689,  0.9691],
        [ 0.0145, -1.9414, -0.2109,  2.6994,  0.7435],
        [ 0.0448,  0.7963, -0.3597,  1.3682,  0.6240],
        [ 0.7535, -0.4397,  1.3501,  1.0288, -0.1718]], device='cuda:0')
=======
tensor([[-0.5169,  0.5627, -0.1216,  1.1110, -0.3380],
        [ 0.9310, -1.2706,  0.9001,  0.7387,  0.2435],
        [ 1.4936, -0.2933,  0.8013, -0.3141, -1.3923],
        [ 0.0598, -0.6556,  0.3131,  0.3376, -1.0293]], device='cuda:0')
tensor([[False, False, False, False, False],
        [False, False, False, False, False],
        [False, False, False, False, False],
        [False, False, False, False, False]], device='cuda:0')
```

### Other

# Environment:
Name: torchsde
Version: 0.2.6

Name: torch
Version: 2.4.0

Python 3.11.9

GPU: NVIDIA GeForce RTX 4090

System: Linux 59561615c495 5.15.0-94-generic #104-Ubuntu SMP Tue Jan 9 15:25:40 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

**Comment by chaObserv** [comment]
It's not a precision issue. The cpu and gpu variants determine where the random noise for SDE is generated. Different devices can produce different random values even with the same seed. ([Reference1](https://discuss.pytorch.org/t/even-with-the-same-seed-different-random-numbers-are-generated-by-different-devices/154054/2), [Reference2](https://stackoverflow.com/a/77473480))

**Comment by caijizhuo** [comment]
Thank you very much! I understand how it works now.
